### ISSUE TITLE
scaffolding existing mature db leaves an endless list of model "corrections" - tooling hell & monkey code ... solutions? 

### ISSUE THREAD

**Issue by wvmstr** [dotnet/efcore#23835]
I have a large mature database that has been paired with EF 6.4 for years. All relationships properly defined, EF 6.4 was happy and allowed me to utilize the benefits of EF. Never ever had to dig into the bowels of EF and I was glad for the black box experience.

In a different (new framework) application I am using the same database for a NET5.0 Web Api together with EF Core 5.01.
I was hoping for the same black box experience, but so far it has been an outright hell with cryptic error messages, endless hours of monkey code , and at least one seemingly unsolvable problem. Wasted weeks of trying different things and made no progress.

This db contains 115 tables, apparently scaffolding cannot determine what the primary keys are in these tables, so I am monkey coding [Key] and "using System.ComponentModel.DataAnnotations;" 115 times over!

Is there a setting I can use so the scaffolding recognizes and put these two items in the correct place for me?

Second, a table called Client has several dependent lookup tables with one-to-many relationships. SSMS shows a nice diagram with keys and many symbols in all the right places, and EF 6.4 was happy! One of the lookup tables is called Gender.

```C#
public partial class EqGender
    {
        public EqGender()
        {
            Clients = new HashSet<Client>();
        }

        public int GId { get; set; }
        public string Gender { get; set; }

        public virtual ICollection<Client> Clients { get; set; }
    }`


public partial class Client
    {
        public Client()
        {
            Appointments = new HashSet<Appointment>();
            ...
            Transactions = new HashSet<Transaction>();
        }

        [Key] public int ClientId { get; set; }
        public string Cname { get; set; }
        public string Cdescription { get; set; }
        public int? OwnerLocationId { get; set; }
        public int? EnteredByProviderId { get; set; }
        public int? OwnerBillToId { get; set; }
        public bool? BillLocation { get; set; }

        public int? BId { get; set; }
        public virtual EqGender GIdNavigation { get; set; }
        
        [ForeignKey("EnteredByProviderId")] public virtual Provider EnteredByProvider { get; set; }
        [ForeignKey("OwnerBillToId")]  public virtual Owner OwnerBillTo { get; set; }
        [ForeignKey("OwnerLocationId")] public virtual Owner OwnerLocation { get; set; }
        
        public virtual ICollection<Appointment> Appointments { get; set; }
        public virtual ICollection<Transaction> Transactions { get; set; }
    }
```

When I query for Client, I get " Microsoft.Data.SqlClient.SqlException (0x80131904): Invalid column name '_BIdNavigationBId_'."

As far as I can see these two Entities are properly setup as per the documentation. What is wrong here?

And I don't even know if this is the last problem I can expect to have to deal with.

**Comment by anranruye** [comment]
it seems that you can still use ef 6.4 for .NET 5. 

and I think it's better to start with your existing ef6 dbcontext to build your ef core dbcontext than using scaffolding.

**Comment by roji** [comment]
@anranruye we generally need to see a very concrete issue, including the database schema being scaffolded and the resulting model.

For your first issue, note that EF Core doesn't scaffold the `[Key]` where that isn't necessary; EF Core automatically detects key properties by convention if they're called `Id` or `<Class>Id` ([see the docs](https://docs.microsoft.com/en-us/ef/core/modeling/keys)). If those are the properties being scaffolded, then there's no need for you to go over it and add the attribute. Otherwise, can you please post a sample table (i.e. a CREATE TABLE statement) which doesn't scaffold correctly for you?

For your second question, we're really lacking the necessary information to investigate. I recommend you open a separate issue (and keep this one for the first problem), and please include a runnable code sample.

**Comment by ErikEJ** [comment]
If you opt in for DataAnnotations, Key annotation is always scaffolded, actually.

**Comment by wvmstr** [comment]
This is what I did in VS2019 16.8.3:
1. create new ASPNet Core Web Application
2. select ASP.NET Core 5.0 and ASP.NET Core Web API (no authentication, configure for HTTPS, Enable OpenAPI support)
3. Browse with HttpRepl and then Get WeatherForecast => works
4. Install EF Core + Tools + SqlServer all version 5.01
5. Scaffold my existing db with: dotnet ef dbcontext scaffold "my connectionstring" Microsoft.EntityFrameworkCore.SqlServer -o Model
6. Models folder get created
7. Create a new controller:
`using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using BtClassLibrary;
using BtApiCore.Model;


namespace BtApiCore.Controllers
{
    [ApiController]
    [Route("api/[controller]")]

    public class ClientController : ControllerBase
    {
        private readonly AppDbContext context;
        public ClientController(AppDbContext context)
        {
            this.context = context;
        }
        [HttpGet]
        public async Task<ActionResult<List<Client>>> Get() => await context.Clients
            //.Include(client => client.EnteredByProvider)
            .ToListAsync();

        [HttpGet("{id}", Name = "GetClient")]
        public async Task<ActionResult<Client>> Get(int id)
        {
            return await context.Clients
                .FirstOrDefaultAsync(x => x.ClientId == id);
        }

    }
}`
8. Add connection string to appsettings.json
9. Add AppDbContext and modify startup.cs with "services.AddDbContext<AppDbContext>(options => options.UseSqlServer(Configuration["ConnectionString:Default"]));"
B
Browse with HttpRepl and  request Get Country

So far so good, all nice out of box functionality, so simple even I can do it (thanks Microsoft!).

500 Internal Server Error:
System.InvalidOperationException: Unable to determine the relationship represented by navigation 'Client.OwnerBillTo' of type 'Owner'.

Here it becomes a night mare. Remember EF 6.4 has looked at these relationships and could work with them, EF Core finds fault and needs Model "customizations". Ok, this relationship is not a cookie cutter one, so with the help of some of my partners in crime at the aspnet forums we figured out how to add these "customizations".
There are 3 more similar relationships and when these are fixed, there is an endless stream of 115 error messages all complaining about it not being clear what the primary keys are for these 115 tables.  
Several hours later I have faithfully copied [Key] in front of the primary keys in all 115 tables, as well as the necessary using statement. In my mind this is tooling hell / monkey code.

Then finally  "get country" returns the expected results.

Then I add ClientController and the whole thing blows up with "Microsoft.Data.SqlClient.SqlException (0x80131904): Invalid column name 'BIdNavigationBId'.""

I do not know how to produce a working code sample, except for a zip file containing the solution and a bak.
Which I can make available to Microsoft.

**Comment by roji** [comment]
@wvmstr there's really nothing we can do without the database schema. If possible, please send a ZIP file to my email address (visible on my github profile).

**Comment by wvmstr** [comment]
> If you opt in for DataAnnotations, Key annotation is always scaffolded, actually.

How can I opt in for DataAnnotations? 

Any thoughts regarding scaffolding from a existing db producing such an endless list of blocking issues? I first ran into this problem middle of December and have not made progress since (5 weeks now).

**Comment by ErikEJ** [comment]
Look at the docs for "dotnet ef dbcontext scaffold"

And in order to help, you must share a repro project.

**Comment by roji** [comment]
Reopening as I've received a repro by mail, and will investigate (though it will take a few days before I have the time).

**Comment by roji** [comment]
This ended up being the result of creating a new DbContext without model configuration rather than using the scaffolded one.

**Comment by ErikEJ** [comment]
Candidate for improved docs??

**Comment by roji** [comment]
IMHO [our doc page on scaffolding](https://docs.microsoft.com/en-us/ef/core/managing-schemas/scaffolding?tabs=dotnet-core-cli) already makes it clear that scaffolding produces a DbContext as well, but we could think about improving...

**Comment by ajcvickers** [comment]
Do we know where the confusion came from?

**Comment by roji** [comment]
@wvmstr feel free to chime in, but as I understand it the user wasn't aware that scaffolding produces a fully-configured DbContext, and created a new one from scratch. This is in the context of an ASP.NET application - it may be a good idea to include some info on scaffolding in that tutorial.

**Comment by ajcvickers** [comment]
@roji Yes, I get that. But _why_ did they think that. For example, if something in the documentation can be interpreted that way, then updating the docs makes sense. On the other hand, the docs might never have been read.

**Comment by roji** [comment]
@wvmstr could we get your input on this?

**Comment by wvmstr** [comment]
Yup, I did indeed create appDbContext after the scaffolding. Using HttpRepl I was able to see that some of my controller actions were indeed producing the desired results, but not all..... at that time I was thinking there were some issues that I did not know how to fix just yet. Posted to ASPNET forums, but that never had a chance without someone looking at my code.
Never went back to a earlier decision point (create appDbContext) and reevaluated my options at that point. The types of exception messages I was getting and the feedback at ASPNET forums were pointing me towards Model customizations. I.e. one rabbit hole after the other.

I do end up on doc.microsoft.com from time to time to find solutions to problems I run into, but I never go there to prepare for an implementation. Usually I get lost in lingo overload, so it is not my first choice. Most of the time I do not know the right terminology to be successful in searching, or the terminology changed to better reflect current thinking (at Microsoft). Looking at [this](https://docs.microsoft.com/en-us/ef/core/managing-schemas/scaffolding?tabs=dotnet-core-cli) again and I must say this page clearly spells out "The scaffolded DbContext class name will be the name of the database suffixed with Context by default."

Just never got to that point.

I do go to YouTube, but I find the examples mostly of the Books/Authors Students/Classes kind. For visual learners a full walk-through is very helpful, almost essential. Not only do these show how it is done, but also the terminology gets highlighted.

I am appreciative of the help I got, and as a result I have made some good progress since, even with that brown paper bag on my head with the word SHAME written on it.
